lua/lvim/updater.lua

```diff
@@ -0,0 +1,233 @@
+      finished = finished + 1
+      print_update()
+    end
+    disp.task_failed = function(_, plugin_name)
+      print("!! " .. plugin_name .. ": failed to install!")
+      finished = finished + 1
+      print_update()
+    end
+    disp.task_start = function()
+      require("packer.display").status.running = true
+      total = total + 1
+      print_update()
+    end
+    return disp
+  end
+end
+
+local function no_op() end
+local treesitter_total = 0
+local treesitter_installed = 0
+
+-- nvim-treesitter has no hook to download asynchronously _and_ wait
+local function inject_treesitter_progress_check()
+  local ts_install_iter_cmd = require("nvim-treesitter.install").iter_cmd
+  local iter_env = setmetatable({}, {
+    __index = function(_, key)
+      -- nvim-treesitter is too noisy, don't let it print
+      if key == "print" then
+        return no_op
+      else
+        return _G[key]
+      end
+    end,
+  })
+  setfenv(ts_install_iter_cmd, iter_env)
+  treesitter_total = 0
+  treesitter_installed = 0
+  local function treesitter_print_update()
+    local line = "Installed " .. treesitter_installed .. " of " .. treesitter_total
+    vim.schedule(function()
+      vim.fn.chansend(1, { "\r" .. line .. "\n" })
+    end)
+  end
+
+  -- Wraps: https://github.com/nvim-treesitter/nvim-treesitter/blob/f88e16ce0d59a49c050fd7efbf37e96293353192/lua/nvim-treesitter/install.lua#L116-L122
+  require("nvim-treesitter.install").iter_cmd = function(cmd_list, i, lang, success_message)
+    if i == 1 then
+      treesitter_total = treesitter_total + 1
+    end
+    if i == #cmd_list + 1 then
+      treesitter_installed = treesitter_installed + 1
+      if not _G.__lvim_test_env then
+        treesitter_print_update()
+      end
+    end
+
+    ts_install_iter_cmd(cmd_list, i, lang, success_message)
+  end
+end
+
+function M.init(opts)
+  if not _G.__lvim_test_env then
+    enable_packer_status()
+  end
+
+  local core_plugins = require "lvim.plugins"
+
+  -- prevent packer from loading plugins on run hooks
+  local load_plugin = require("packer.plugin_utils").load_plugin
+  local post_update_hook = require("packer.plugin_utils").post_update_hook
+  require("packer.plugin_utils").load_plugin = function() end
+
+  local packer_stage = 0
+
+  async(function()
+    -- enable all core plugins initially until after the user config is loaded
+    for _, plug in ipairs(core_plugins) do
+      plug.disable = false
+    end
+
+    if opts.packer_cache_path then
+      vim.fn.delete(opts.packer_cache_path)
+    end
+
+    local plugin_loader = require "lvim.plugin-loader"
+    plugin_loader.load { core_plugins }
+
+    local user_lvim
+    packer.on_complete = function() end
+    packer.on_compile_done = function()
+      if packer_stage == 0 then
+        print("\rInstalled core plugins in:", timer:stop(), "ms")
+        packer_stage = 1
+
+        timer:start()
+        print "Loading core plugins..."
+        packer.compile()
```

Comment:
You need to be careful with this, since it's async and it might actually be slower than you think.

Reply:
No, not really.

This entire section of code handles the asynchronous callbacks of Packer in stages and it is in fact not slow. I'm well aware of how the internals of Packer work.